README.md
# Introduction

This Inkscape extension searches all text elements in a document and renders
them with PdfLatex. The rendered text will appear at the same position as the
original text and may contain Latex commands and math formula. Modifications
of the source text will be considered when the extension is executed again.

As the rendered text is added to a new layer, the original document content
is not touched and you can always wipe out all modifications from this extension
by simply deleting the created layer.

# Installation

### Linux

Required packages on Debian-based distributions:

  * `inkscape` (version >= 0.91)
  * `texlive-base` (provides pdflatex, ...)
  * `pdf2svg`
  * `python` and `python-lxml` (Python version < 3.0)
  * `python-gi`, the PyGObject bindings for the GTK3 GUI variant

In other distributions the packages may be named differently, e.g the PyGObject 
bindings on OpenSuse are in the `python2-gobject-Gdk` package.

When the dependencies are installed, just run the `install.sh` script to copy the 
extension into the Inkscape extension folder in the current user home 
directory (`~/.config/inkscape/extensions/`).

### Windows

Required Software:

   * Inkscape (version >= 0.92.2)
   * MiKTeX 2.9

Simply download the installer package from the
[latest release](https://github.com/seebk/LaTeXText/releases) and run it.

__NOTE:__ You need at least Inkscape 0.92.2. Prior versions used a different
build environment on Windows and are not compatible with this extension.

### Mac OS X

Required Software:

   * Inkscape (version >= 0.91)
   * MacTex
   * pdf2svg (e.g. installed from Macports)
   * Python lxml module
   * Python PyGObject module for the GTK3 GUI variant

Instructions:

  1. Open a command terminal and check if you can execute `pdflatex`. If it is not
  found verify if MacTex is up-to-date and properly installed.

  2. Install pdf2svg from [Macports](https://www.macports.org/).

         $ sudo port install pdf2svg

  3. Install the required Python modules.

         $ sudo port install py27-gobject3
         $ sudo port install py27-lxml

  4. Macports does not install the Python modules into the default Apple Python
     environment and the official Inkscape OS X installer will not use Python from
     Macports. So we have to install Inkscape from Macports, too.

         $ sudo port install inkscape

     As this pulls quite a lot of dependencies, the download and compilation may
     take a long time. Take a coffee ;)

  5. Run the `install.sh` script to copy the extension into the Inkscape extension
  folder in the current user home directory (`~/.config/inkscape/extensions/`).

If there is an easier way to install and use the extension on OS X, in particular
in combination with the official Inkscape installer, feel free to send updated
instructions.


# Usage

### Inkscape extension

![Step-by-step animation](https://media.giphy.com/media/3ov9jG1BQAWY9LAOHu/source.gif)

General instructions:

  * Put all text elements on a separate layer.

  * Run 'Extensions -> Render -> Text with Latex (GTK3)'. It will create a new
  layer with all text elements being rendered by `pdflatex`.

  * Switch visibility of the layers to either see the Latex code or to see the
  rendered result.

  * If you changed your Latex code or added new text elements, simply re-run
  the extension and it will update the render layer.

__NOTE:__ There may be two different entries in the Inkscape extension menu: a standard
Inkscape extension and a 'GTK3' GUI variant. The latter is the recommended one
as it can remember and restore previous settings per document and has a more
comfortable UI. The other extension is only there to be used as a fallback if
GTK3 is not available.

#### Options ####

  * _Preamble File_ -- A Latex preamble file to load and configure additional
    packages. The `\documentclass` and `\begin{document}` should not be included,
    see the `preambles/` subfolder for examples.
    
  * _Additional Packages_ -- A comma separated list of additional Latex packages 
    that should be loaded.

  * _Document base font size_ -- The main font size for each text element. This
    will correspond to the font size otherwise defined in the `\documentclass`
    statement of a Latex document. Relative font size changes can be set in the
    text with the standard Latex commands (`\small`, `\tiny`, `\Large`, ...).

  * _Scale factor_ -- An additional scaling applied to each rendered element.

  * _SVG/XML tree max. depth_ -- Some imported graphics may have deeply nested
    text elements we do not want to render. If this option is greater than zero
    text elements above the given depth will be ignored.

  * _Add `\\` at every line break_ -- Automatically replace each line break in a
    text element by a Latex line break (`\\`)

  * _Encapsulate all text with $..$_ -- Put all text in math mode by default.

  * _Show log messages_ -- Show log messages for debugging purpose (if there is
    any Latex error the log will be shown anyway)


### Commandline mode

You can directly run the extension in the commandline outside of Inkscape, too.

    Usage: latextext.py [options] SVGfile(s)

    Options:
    -h, --help            show this help message and exit
    -o FILE, --outfile=FILE
                        write to output file or directory
    -p FILE, --preamble=FILE
                        latex preamble file
    -f FONTSIZE, --fontsize=FONTSIZE
                        latex base font size
    -s SCALE, --scale=SCALE
                        apply additional scaling
    -d DEPTH, --depth=DEPTH
                        maximum search depth for grouped text elements
    -n, --newline         insert  ewline at every line break
    -m, --math            encapsulate all text in math mode
    -c, --clean           remove all renderings
    -v, --verbose      


# Credits

Some of the code is based on the [textext](https://pav.iki.fi/software/textext/)
extension from Pauli Virtanen. 

# License

This software is licensed under the GPLv3 license. A copy of the license can
be found in `LICENSE.txt`


CHANGES.md

### 2019-03-09 - v0.1.2 ###

* New option to directly enter a list of additional Latex packages to be 
  loaded without using a preamble file.

* Win64 Installer: reduce size and only install the GTK DLLs that are 
  absolutely required.


### 2017-11-13 - v0.1.1 ###

Fixed two problems with the Windows installer.

    * Add the Adwaita icons.

    * Add a gschemas.compiled file to avoid crash when the preamble file 
      chooser is openend.


### 2017-10-30 - v0.1 ###

Initial release.

install.sh
#!/bin/bash

DEPENDENCIES=true
GTK3=false

echo -n "Check if Python LXML is available... "
python -c "import lxml" 2> /dev/null
if [ $? -ne 0 ]; then
    echo "NOT FOUND"
    DEPENDENCIES=false
else
    echo "OK"
fi

if [ "$DEPENDENCIES" = false ]; then
    echo "ERROR: some required Python modules are missing."
    exit
fi

echo -n "Check if PyGObject is available... "
python -c "import gi" 2> /dev/null
if [ $? -ne 0 ]; then
    echo "NOT FOUND, installing standard extension GUI"
    GTK3=false
else
    GTK3=true
    echo "OK"
fi

echo -n "Copying extension to ~/.config/inkscape/extensions/ ... "

cp extension/latextext.py ~/.config/inkscape/extensions/
    
if [ "$GTK3" = true ]; then
    cp extension/latextext_gtk3.py ~/.config/inkscape/extensions/
    cp extension/latextext_gtk3.inx ~/.config/inkscape/extensions/
else
    cp extension/latextext.inx ~/.config/inkscape/extensions/
fi

echo "done"


render_latex_gtk3.py
#!/usr/bin/env python

######################
#  Check if we can create a nice GTK3 GUI
try:
    import gi
    gi.require_version('Gtk', '3.0')
    from gi.repository import Gtk, Gdk
    GTK3_AVAILABLE = True
except ImportError:
    GTK3_AVAILABLE = False
    print("PyGObject module was not found!")
    raise

######################
#  Check if we are in the inkscape extension folder
try:
    import inkex
except ImportError:
    print("GTK3 variant has to be run as an Inkscape extension!")
    raise

from render_latex import RenderLatexEffect, SvgProcessor, set_log_level, log_level_debug


######################
# GTK Dialog
#    Extended GTK3 user interface as a replacement of the simple Inkscape
#    default interface
class Gtk3ParamGui(Gtk.Window):

    def __init__(self, run_callback):
        self.run_callback = run_callback
        Gtk.Window.__init__(self, title="Render Latex Layer (GTK3)")
        gh = Gdk.Geometry()
        gh.max_height = 120
        gh.max_width = 700
        self.set_geometry_hints(None, gh, Gdk.WindowHints.MAX_SIZE)
        self.set_hexpand(True)
        self.set_size_request(400, gh.max_height)
        self.set_keep_above(True)

    def prepare_dialog(self, options):

        self.options = options

        grid = Gtk.Grid()
        grid.set_row_spacing(3)
        grid.set_column_spacing(10)
        grid.set_hexpand(True)
        grid.set_border_width(10)
        self.add(grid)

        row_count = 0
        box0 = Gtk.Box(spacing=6)
        grid.attach(Gtk.Label("Preamble File"), 0, row_count, 1, 1)
        self.entryPreamble = Gtk.Entry()
        self.entryPreamble.set_text(options.preamble)
        self.entryPreamble.set_hexpand(True)
        box0.pack_start(self.entryPreamble, True, True, 0)
        btnSelectPreamble = Gtk.Button(label="...")
        btnSelectPreamble.set_hexpand(False)
        btnSelectPreamble.connect("clicked", self.on_select_preamble)
        box0.pack_start(btnSelectPreamble, False, False, 0)
        grid.attach(box0, 1, row_count, 1, 1)

        row_count += 1
        grid.attach(Gtk.Label("Document base font size"), 0, row_count, 1, 1)
        self.entryFontsize = Gtk.SpinButton.new_with_range(1, 32, 1)
        self.entryFontsize.set_value(options.fontsize)
        self.entryFontsize.set_hexpand(False)
        grid.attach(self.entryFontsize, 1, row_count, 1, 1)

        row_count += 1
        self.entryScale = Gtk.SpinButton.new_with_range(0.01, 20.0, 0.05)
        self.entryScale.set_digits(2)
        self.entryScale.set_value(options.scale)
        grid.attach(Gtk.Label("Scale factor"), 0, row_count, 1, 1)
        grid.attach(self.entryScale, 1, row_count, 1, 1)

        row_count += 1
        self.entryDepth = Gtk.SpinButton.new_with_range(0, 100, 1)
        self.entryDepth.set_value(options.depth)
        grid.attach(Gtk.Label("SVG/XML tree max. depth"), 0, row_count, 1, 1)
        grid.attach(self.entryDepth, 1, row_count, 1, 1)

        row_count += 1
        grid.attach(Gtk.Label("Add \\\\ at every line break"), 0, row_count, 1, 1)
        self.btnNewline = Gtk.CheckButton()
        self.btnNewline.set_active(options.newline)
        grid.attach(self.btnNewline, 1, row_count, 1, 1)

        row_count += 1
        grid.attach(Gtk.Label("Encapsulate all text with ($..$)"), 0, row_count, 1, 1)
        self.btnMath = Gtk.CheckButton()
        self.btnMath.set_active(options.math)
        grid.attach(self.btnMath, 1, row_count, 1, 1)

        row_count += 1
        grid.attach(Gtk.Label("Show log messages"), 0, row_count, 1, 1)
        self.btnShowLog = Gtk.CheckButton()
        grid.attach(self.btnShowLog, 1, row_count, 1, 1)

        row_count += 2
        box1 = Gtk.Box(spacing=6)
        grid.attach(box1, 0, row_count, 2, 1)

        btnApply = Gtk.Button(label="Apply")
        btnApply.connect("clicked", self.on_btnApply_clicked)
        box1.pack_end(btnApply, False, False, 0)

        btnClose = Gtk.Button(label="Close")
        btnClose.connect("clicked", self.on_btnClose_clicked)
        box1.pack_end(btnClose, False, False, 0)

        self.show_all()
        self.entryScale.grab_focus()
        btnApply.set_can_default(True)
        btnApply.grab_default()
        btnApply.grab_focus()
        self.set_default(btnApply)

    def on_select_preamble(self, widget):
        dialog = Gtk.FileChooserDialog("Please choose a Latex preamble file", self,
                                       Gtk.FileChooserAction.OPEN,
                                       (Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL,
                                        Gtk.STOCK_OPEN, Gtk.ResponseType.OK))

        response = dialog.run()
        if response == Gtk.ResponseType.OK:
            self.entryPreamble.set_text(dialog.get_filename())

        dialog.destroy()

    def on_btnClose_clicked(self, widget):
        Gtk.main_quit()

    def on_btnApply_clicked(self, widget):
        # get parameters
        self.options.scale = self.entryScale.get_value()
        self.options.depth = self.entryDepth.get_value()
        self.options.fontsize = self.entryFontsize.get_value()
        self.options.preamble = self.entryPreamble.get_text()
        self.options.math = self.btnMath.get_active()
        self.options.newline = self.btnNewline.get_active()
        if self.btnShowLog.get_active() is True:
            set_log_level(log_level_debug)
        try:
            self.run_callback()
        finally:
            Gtk.main_quit()


# Extend standard inkscape extension with the GTK3 GUI
class RenderLatexEffectGTK3(RenderLatexEffect):

    def effect(self):
        svgprocessor = SvgProcessor(self.document, self.options)

        ParamGui = Gtk3ParamGui(svgprocessor.run)
        ParamGui.connect("delete-event", Gtk.main_quit)
        ParamGui.prepare_dialog(svgprocessor.options)
        Gtk.main()


if __name__ == "__main__":
    effect = RenderLatexEffectGTK3()
    effect.affect()


tools/innosetup/prepare.ps1

# Download ressources from MSYS2
$msys2_repo= "http://repo.msys2.org/mingw/x86_64/"

$client = new-object System.Net.WebClient
$zipExe = join-path ${env:ProgramFiles} '7-zip\7z.exe'
$pwd = Get-Location

function get_msys2_package($pkg){	
    If (-Not (Test-Path "$pkg")){
    	Write-Host "Downloading $pkg ..."
        $client.DownloadFile("$msys2_repo/$pkg", (Join-Path $pwd $pkg))
    }
    Write-Host "Extracting $pkg ..."
    &$zipExe x $pkg -y | Out-Null
    &$zipExe x $pkg.Substring(0,$pkg.Length-3) -y -opygobject | Out-Null
    Remove-Item $pkg.Substring(0,$pkg.Length-3)
}

get_msys2_package("mingw-w64-x86_64-atk-2.24.0-1-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-cairo-1.15.4-4-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-gdk-pixbuf2-2.36.6-2-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-gobject-introspection-runtime-1.52.0-0-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-gtk3-3.22.15-1-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-libepoxy-1.4.2-1-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-pango-1.40.6-1-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-python2-cairo-1.13.3-2-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-python2-gobject-3.24.1-1-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-adwaita-icon-theme-3.24.0-1-any.pkg.tar.xz")

# Dependencies for glib-compile-schemas.exe
get_msys2_package("mingw-w64-x86_64-glib2-2.52.0-1-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-pcre-8.41-1-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-gettext-0.19.8.1-3-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-libwinpthread-git-5.0.0.4833.f057c525-1-any.pkg.tar.xz")
get_msys2_package("mingw-w64-x86_64-libiconv-1.15-1-any.pkg.tar.xz")


# Compile gschemas
Write-Host "Compiling glib schemas ..."
&"pygobject\mingw64\bin\glib-compile-schemas.exe" pygobject\mingw64\share\glib-2.0\schemas

# pdf2svg
If (-Not (Test-Path "pdf2svg.zip")){
	Write-Host "Downloading pdf2svg ..."
    $client.DownloadFile("https://github.com/jalios/pdf2svg-windows/archive/master.zip", "pdf2svg.zip")
}
Write-Host "Extracting pdf2svg ..."
Expand-Archive pdf2svg.zip -Force -dest .

extension/latextext_gtk3.py
#!/usr/bin/env python

######################
#  Check if we can create a nice GTK3 GUI
try:
    import gi
    gi.require_version('Gtk', '3.0')
    from gi.repository import Gtk, Gdk
    GTK3_AVAILABLE = True
except ImportError:
    GTK3_AVAILABLE = False
    print("PyGObject module was not found!")
    raise

######################
#  Check if we are in the inkscape extension folder
try:
    import inkex
except ImportError:
    print("GTK3 variant has to be run as an Inkscape extension!")
    raise

from latextext import RenderLatexEffect, SvgProcessor, set_log_level, log_level_debug


######################
# GTK Dialog
#    Extended GTK3 user interface as a replacement of the simple Inkscape
#    default interface
class Gtk3ParamGui(Gtk.Window):

    def __init__(self, run_callback):
        self.run_callback = run_callback
        Gtk.Window.__init__(self, title="LaTexText (GTK3)")
        gh = Gdk.Geometry()
        gh.max_height = 120
        gh.max_width = 700
        self.set_geometry_hints(None, gh, Gdk.WindowHints.MAX_SIZE)
        self.set_hexpand(True)
        self.set_size_request(400, gh.max_height)
        self.set_keep_above(True)

    def prepare_dialog(self, options):

        self.options = options

        grid = Gtk.Grid()
        grid.set_row_spacing(3)
        grid.set_column_spacing(10)
        grid.set_hexpand(True)
        grid.set_border_width(10)
        self.add(grid)

        row_count = 0
        box0 = Gtk.Box(spacing=6)
        grid.attach(Gtk.Label("Preamble File"), 0, row_count, 1, 1)
        self.entryPreamble = Gtk.Entry()
        self.entryPreamble.set_text(options.preamble)
        self.entryPreamble.set_hexpand(True)
        box0.pack_start(self.entryPreamble, True, True, 0)
        btnSelectPreamble = Gtk.Button(label="...")
        btnSelectPreamble.set_hexpand(False)
        btnSelectPreamble.connect("clicked", self.on_select_preamble)
        box0.pack_start(btnSelectPreamble, False, False, 0)
        grid.attach(box0, 1, row_count, 1, 1)

        row_count += 1
        grid.attach(Gtk.Label("Additional Packages"), 0, row_count, 1, 1)
        self.entryPackages = Gtk.Entry()
        self.entryPackages.set_text(options.packages)
        self.entryPackages.set_hexpand(True)
        grid.attach(self.entryPackages, 1, row_count, 1, 1)

        row_count += 1
        grid.attach(Gtk.Label("Document base font size"), 0, row_count, 1, 1)
        self.entryFontsize = Gtk.SpinButton.new_with_range(1, 32, 1)
        self.entryFontsize.set_value(options.fontsize)
        self.entryFontsize.set_hexpand(False)
        grid.attach(self.entryFontsize, 1, row_count, 1, 1)

        row_count += 1
        self.entryScale = Gtk.SpinButton.new_with_range(0.01, 20.0, 0.05)
        self.entryScale.set_digits(2)
        self.entryScale.set_value(options.scale)
        grid.attach(Gtk.Label("Scale factor"), 0, row_count, 1, 1)
        grid.attach(self.entryScale, 1, row_count, 1, 1)

        row_count += 1
        self.entryDepth = Gtk.SpinButton.new_with_range(0, 100, 1)
        self.entryDepth.set_value(options.depth)
        grid.attach(Gtk.Label("SVG/XML tree max. depth"), 0, row_count, 1, 1)
        grid.attach(self.entryDepth, 1, row_count, 1, 1)

        layer_list = Gtk.ListStore(str, str)
        layer_list.append(['All layers','all'])
        active_index = 0
        i = 0
        for layer in self.options.layers:
            i += 1
            if self.options.active_layer == layer[0]:
                active_index = i
            layer_list.append([layer[1], layer[0]])

        row_count += 1
        grid.attach(Gtk.Label("Selected layer"), 0, row_count, 1, 1)
        self.comboLayer = Gtk.ComboBox.new_with_model(layer_list)
        renderer_text = Gtk.CellRendererText()
        self.comboLayer.pack_start(renderer_text, True)
        self.comboLayer.add_attribute(renderer_text, "text", 0)
        self.comboLayer.set_active(active_index)
        grid.attach(self.comboLayer, 1, row_count, 1, 1)

        row_count += 1
        grid.attach(Gtk.Label("Add \\\\ at every line break"), 0, row_count, 1, 1)
        self.btnNewline = Gtk.CheckButton()
        self.btnNewline.set_active(options.newline)
        grid.attach(self.btnNewline, 1, row_count, 1, 1)

        row_count += 1
        grid.attach(Gtk.Label("Encapsulate all text with ($..$)"), 0, row_count, 1, 1)
        self.btnMath = Gtk.CheckButton()
        self.btnMath.set_active(options.math)
        grid.attach(self.btnMath, 1, row_count, 1, 1)

        row_count += 1
        grid.attach(Gtk.Label("Show log messages"), 0, row_count, 1, 1)
        self.btnShowLog = Gtk.CheckButton()
        grid.attach(self.btnShowLog, 1, row_count, 1, 1)

        row_count += 2
        box1 = Gtk.Box(spacing=6)
        grid.attach(box1, 0, row_count, 2, 1)

        btnApply = Gtk.Button(label="Apply")
        btnApply.connect("clicked", self.on_btnApply_clicked)
        box1.pack_end(btnApply, False, False, 0)

        btnClose = Gtk.Button(label="Close")
        btnClose.connect("clicked", self.on_btnClose_clicked)
        box1.pack_end(btnClose, False, False, 0)

        self.show_all()
        self.entryScale.grab_focus()
        btnApply.set_can_default(True)
        btnApply.grab_default()
        btnApply.grab_focus()
        self.set_default(btnApply)

    def on_select_preamble(self, widget):
        dialog = Gtk.FileChooserDialog("Please choose a Latex preamble file", self,
                                       Gtk.FileChooserAction.OPEN,
                                       (Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL,
                                        Gtk.STOCK_OPEN, Gtk.ResponseType.OK))
        dialog.set_keep_above(True)
        dialog.set_modal(True)
        response = dialog.run()
        if response == Gtk.ResponseType.OK:
            self.entryPreamble.set_text(dialog.get_filename())

        dialog.destroy()

    def on_btnClose_clicked(self, widget):
        Gtk.main_quit()

    def on_btnApply_clicked(self, widget):
        # get parameters
        self.options.scale = self.entryScale.get_value()
        self.options.depth = self.entryDepth.get_value()
        self.options.fontsize = self.entryFontsize.get_value()
        self.options.preamble = self.entryPreamble.get_text()
        self.options.packages = self.entryPackages.get_text()
        self.options.math = self.btnMath.get_active()
        self.options.newline = self.btnNewline.get_active()
        self.options.active_layer = self.comboLayer.get_model()[self.comboLayer.get_active_iter()][1]
        if self.btnShowLog.get_active() is True:
            set_log_level(log_level_debug)
        try:
            self.run_callback()
        finally:
            Gtk.main_quit()


# Extend standard inkscape extension with the GTK3 GUI
class RenderLatexEffectGTK3(RenderLatexEffect):

    def effect(self):
        svgprocessor = SvgProcessor(self.document, self.options)

        ParamGui = Gtk3ParamGui(svgprocessor.run)
        ParamGui.connect("delete-event", Gtk.main_quit)
        ParamGui.prepare_dialog(svgprocessor.options)
        Gtk.main()


if __name__ == "__main__":
    effect = RenderLatexEffectGTK3()
    effect.affect()
